great_expectations/zep/sources.py

```diff
@@ -40,84 +41,180 @@ class _SourceFactories:
+            raise TypeRegistrationError(
+                f"'{ds_type_name}' - `sources.{method_name}()` factory already exists",
+                field_name="type",
+                type_=ds_type,
+            )
 
-        # TODO: simplify or extract the following datasource & asset type registration logic
-        asset_type_names = [
-            _get_simplified_name_from_type(t, suffix_to_remove="_asset")
-            for t in asset_types
-        ]
+        datasource_type_lookup[ds_type] = ds_type_name
+        LOGGER.info(f"'{ds_type_name}' added to `type_lookup`")
+        cls.__source_factories[method_name] = factory_fn
+        return ds_type_name
 
-        # TODO: We should namespace the asset type to the datasource so different datasources can reuse asset types.
-        already_registered_assets = set(asset_type_names).intersection(
-            cls.type_lookup.keys()
-        )
-        if already_registered_assets:
-            raise ValueError(
-                f"The following names already have a registered type - {already_registered_assets} "
+    @classmethod
+    def _register_engine(
+        cls,
+        ds_type: Type[Datasource],
+        type_lookup_name: str,
+        engine_type_lookup: TypeLookup,
+    ):
+        try:
+            exec_engine_type: Type[ExecutionEngine] = ds_type.__fields__[
+                "execution_engine"
+            ].type_
+        except (AttributeError, KeyError) as exc:
+            LOGGER.warning(f"{exc.__class__.__name__}:{exc}")
+            raise TypeError(
+                f"No `execution_engine` found for {ds_type.__name__} unable to register `ExecutionEngine` type"
+            ) from exc
+
+        eng_class_name: str = exec_engine_type.__name__
+        if eng_class_name == "ExecutionEngine":
+            raise TypeRegistrationError(
+                f"`{ds_type.__name__}.execution_engine` must be annotated with a concrete `ExecutionEngine`",
+                type_=ds_type,
+                field_name="execution_engine",
             )
 
-        for type_, name in zip(asset_types, asset_type_names):
-            cls.type_lookup[type_] = name
-            LOGGER.debug(f"'{name}' added to `type_lookup`")
+        LOGGER.info(
+            f"2c. Registering `ExecutionEngine` type `{eng_class_name}` for '{type_lookup_name}'"
+        )
+        engine_type_lookup[type_lookup_name] = exec_engine_type
+        LOGGER.info(list(engine_type_lookup.keys()))
 
-        cls.type_lookup[ds_type] = simplified_name
-        LOGGER.debug(f"'{simplified_name}' added to `type_lookup`")
-        cls.__source_factories[method_name] = fn
+    @classmethod
+    def _register_assets(cls, ds_type: Type[Datasource], asset_type_lookup: TypeLookup):
+        errored_on: Optional[Type[DataAsset]] = None
+        try:
+            asset_types: List[Type[DataAsset]] = ds_type.asset_types
+
+            if not asset_types:
+                LOGGER.warning(
```

Comment:
Probably, this is a bit of laziness on my part because I initially wanted to be able to define minimal `Datasources` without assets just to show the metaclass magic.